### PR TITLE
Hotfix: Process async events by awaiting them in order

### PR DESCRIPTION
**Pull request by reyhanlioglu**
## Summary by CodeRabbit

- **New Features**
  - Enhanced event handling mechanism for improved processing and control flow.
  - Introduced a new method for sequential event processing with filtering criteria.

- **Bug Fixes**
  - Improved robustness of event processing to prevent overlapping calls.

- **Refactor**
  - Updated event management strategy, replacing the previous event bus with a list for better control.
  - Added a mutex for managing concurrent access during event processing.

- **Chores**
  - Added a new dependency for managing mutual exclusion in concurrent programming.

### CODE REVIEW

File: lib/core/widgets/neo_transition_listener/bloc/neo_transition_listener_bloc.dart [hunk 1]
```diff
@@ -93,42 +94,58 @@ class NeoTransitionListenerBloc extends Bloc<NeoTransitionListenerEvent, NeoTran
...
       }
-    });
+
+      for (final event in processedEvents) {
+        _eventList.remove(event);
+      }
+
+      _isProcessing = false;
+      await _processEvents();
+    }
```
[the remark this answers is not in the document]
Comment (reyhanlioglu) [reply lib/core/widgets/neo_transition_listener/bloc/neo_transition_listener_bloc.dart:148]: Unfortunately it needs to be recursive

File: lib/core/widgets/neo_transition_listener/bloc/neo_transition_listener_bloc.dart [hunk 2]
```diff
@@ -93,42 +94,58 @@ class NeoTransitionListenerBloc extends Bloc<NeoTransitionListenerEvent, NeoTran
...
-        if (_postTransitionTimeoutCompleter != null && !_postTransitionTimeoutCompleter!.isCompleted) {
-          _postTransitionTimeoutTimer?.cancel();
-          _postTransitionTimeoutCompleter?.complete();
+  Future<void> _processEvents() async {
+    if (_isProcessing) {
+      return;
+    }
+    if (_eventList.isNotEmpty) {
+      _isProcessing = true;
+      _eventList.sortBy((element) => element.transition.time);
```
[the remark this answers is not in the document]
Comment (reyhanlioglu) [reply lib/core/widgets/neo_transition_listener/bloc/neo_transition_listener_bloc.dart:105]: all events must have time property, it's okay

File: lib/core/widgets/neo_transition_listener/bloc/neo_transition_listener_bloc.dart [hunk 3]
```diff
@@ -345,10 +362,12 @@ class NeoTransitionListenerBloc extends Bloc<NeoTransitionListenerEvent, NeoTran
...
   void _addEventToBus(NeoSignalREvent event) {
-    if (!_eventBus.values.contains(event)) {
-      if (!isClosed && !_eventBus.isClosed && _lastProcessedEvent?.eventId != event.eventId) {
-        _eventBus.add(event);
+    if (!_eventList
+        .any((element) => element.eventId == event.eventId || element.transition.time.isAfter(event.transition.time))) {
+      if (!isClosed && _lastProcessedEvent?.eventId != event.eventId) {
+        _eventList.add(event);
       }
+      _processEvents();
```
[the remark this answers is not in the document]
Comment (reyhanlioglu) [reply lib/core/widgets/neo_transition_listener/bloc/neo_transition_listener_bloc.dart:370]: there is not any problem in this condition, events must have uniqueue eventId and time must be after than any processed event

File: lib/core/widgets/neo_transition_listener/bloc/neo_transition_listener_bloc.dart [hunk 4]
```diff
@@ -93,42 +94,58 @@ class NeoTransitionListenerBloc extends Bloc<NeoTransitionListenerEvent, NeoTran
...
+        processedEvents.add(event);
       }
-    });
+
+      for (final event in processedEvents) {
+        _eventList.remove(event);
+      }
+
+      _isProcessing = false;
+      await _processEvents();
```
[the remark this answers is not in the document]
Comment (reyhanlioglu) [reply lib/core/widgets/neo_transition_listener/bloc/neo_transition_listener_bloc.dart:147]: what is the suggested code for it?

File: lib/core/widgets/neo_transition_listener/bloc/neo_transition_listener_bloc.dart [hunk 5]
```diff
@@ -93,42 +94,58 @@ class NeoTransitionListenerBloc extends Bloc<NeoTransitionListenerEvent, NeoTran
...
+      _eventList.sortBy((element) => element.transition.time);
+      final List<NeoSignalREvent> processedEvents = [];
+
+      for (final event in List<NeoSignalREvent>.from(_eventList)) {
+        if (isClosed ||
+            event.transition.instanceId.isEmpty ||
+            !(event.transition.instanceId == neoWorkflowManager.instanceId ||
+                event.transition.instanceId == neoWorkflowManager.subFlowInstanceId)) {
+          processedEvents.add(event);
+          break;
```
[the remark this answers is not in the document]
Comment (reyhanlioglu) [reply lib/core/widgets/neo_transition_listener/bloc/neo_transition_listener_bloc.dart]: done

File: lib/core/widgets/neo_transition_listener/bloc/neo_transition_listener_bloc.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (reyhanlioglu) [reply lib/core/widgets/neo_transition_listener/bloc/neo_transition_listener_bloc.dart]: isn't it _isProcessing variable unnecessary ?

[the remark this answers is not in the document]
Comment (reyhanlioglu) [reply lib/core/widgets/neo_transition_listener/bloc/neo_transition_listener_bloc.dart]: calling _processEvents in _processEvents causes deadlock

[the remark this answers is not in the document]
Comment (reyhanlioglu) [reply lib/core/widgets/neo_transition_listener/bloc/neo_transition_listener_bloc.dart]: refactored

**Review by ayhncvdr2 — APPROVED** [review]
LGTM!

**Pull request merged**